content/blog/2023/2023-12-13-high-order-functions-and-inline-functions.md

```diff
@@ -0,0 +1,79 @@
+---
+title: "Exploring the Power of Functions in Kotlin: High-Order Functions vs. Inline Functions"
+categories: ["Kotlin"]
+date: 2023-12-13 00:00:00 +1100
+authors: [ezra]
+excerpt: "In this tutorial, we'll discuss what high order functions and inline functions are in kotlin"
+image: images/stock/0104-on-off-1200x628-branded.jpg
+url: inline functions and high order functions in kotlin
+---
+
+One of Kotlin's standout features is its robust support for functions, including high-order functions and inline functions. In this blog post, we will delve into the world of functions in Kotlin, exploring the differences between high-order functions and inline functions and understanding when to leverage each for optimal code efficiency.
+
+## High-Order Functions
+
+High-order functions are a fundamental concept in functional programming and play a crucial role in Kotlin's functional paradigm. A high-order function is a function that takes one or more functions as parameters or returns a function. This ability to treat functions as first-class citizens allows for more modular and flexible code.
+
+The primary advantage of high-order functions lies in their ability to promote code reuse. By passing functions as parameters, developers can create generic functions that operate on various data types or behaviors. This enhances the readability and maintainability of the code by isolating specific functionalities into separate functions.
+
+Consider the following example of a simple high-order function in Kotlin:
+```kotlin
+fun <T> List<T>.customFilter(predicate: (T) -> Boolean): List<T> {
```

Comment:
After reading the section about inline functions (thanks, btw, I didn't know about them before), I'm asking myself if I can't just add the `inline` keyword in front of this function here as well and get the performance benefit of inline functions. 

What I'm saying is that from this article I don't understand the difference between inline and high-order functions enough:
- why can't we just make all functions `inline`?
- isn't the `inline` function you show in the code snippet below "just" a high-order function with an additional `inline` keyword?

Can you elaborate on these questions in the section "When to use ..." that I added? Please pull the changes, as I pushed a commit.

Reply:
Updated this in the section "When to use....."